### DISCUSSION TITLE
Emacs freezes randomly (eglot + haskell-language-server)

### DISCUSSION THREAD

**Discussion by VictorCMiraldo** [joaotavora/eglot#1151] [Q&A]
Recently, my emacs started randomly freezing when writing Haskell code with eglot and HLS. The only way out is to `pkill -SIGUSR2 emacs` and I get the following trace consistently:
 
```
Debugger entered--entering a function:
* substitute-command-keys("Quit")
  cl-mapcar(markdown-find-previous-prop (markdown-tilde-fence-begin markdown-yaml-metadata-begin markdown-gfm-block-begin))
  markdown-code-block-lang()
  markdown-fontify-code-blocks-generic(markdown-match-gfm-code-blocks 1009511)
  markdown-fontify-gfm-code-blocks(1009511)
  font-lock-fontify-keywords-region(1 1009511 nil)
  font-lock-default-fontify-region(1 1009511 nil)
  font-lock-fontify-region(1 1009511)
  #f(compiled-function (beg end) #<bytecode -0x177a1e0d0ea6dd87>)(1 1009511)
  font-lock-ensure()
  eglot--format-markup((:kind "markdown" :value "\n```haskell\n$dShow :: Show DownloadAttempt\n```\n\n* ..."))
  eglot--hover-info((:kind "markdown" :value "\n```haskell\n$dShow :: Show DownloadAttempt\n```\n\n* ...") (:end (:character 23 :line 516) :start (:character 4 :line 14)))
  #f(compiled-function (jsonrpc-lambda-elem12) #<bytecode -0x118e682f58fbd703>)((:contents (:kind "markdown" :value "\n```haskell\n$dShow :: Show DownloadAttempt\n```\n\n* ...") :range (:end (:character 23 :line 516) :start (:character 4 :line 14))))
  jsonrpc-connection-receive(#<eglot-lsp-server eglot-lsp-server-19ae71c> (:id 34 :jsonrpc "2.0" :result (:contents (:kind "markdown" :value "\n```haskell\n$dShow :: Show DownloadAttempt\n```\n\n* ...") :range (:end (:character 23 :line 516) :start (:character 4 :line 14)))))
  jsonrpc--process-filter(#<process EGLOT (company/haskell-mode)> "           :+: (C1 ('MetaCons \\\"CurlLoginDenied\\\" ...")
```

My current solution is to redefine `eglot--format-markup` to return the empty string, and that stops the problem, but also stops the handy doumentation from showing up.

```
(defun eglot--format-markup (markup) "")
```

Any ideas for pinpointing this bug a little further and solving this in a more principle way? Unfortunately, I cannot work on a MRE right now.

edit 1:

I start to suspect this is a problem in the haskell-language-server, which sends a ridiculous amount of markup and emacs hangs processing it. I'm using:
`haskell-language-server version: 1.8.0.0 (GHC: 9.2.4)` and defining `eglot--format-markup` to `(message "%s" markup)` will make emacs hang a little and display way too much info on the messages buffer.

That being said, I still believe `eglot` should not cause emacs to hang due to a bad language server.

**Comment by manuel-uberti** [comment]
Somewhat related: https://github.com/joaotavora/eglot/discussions/1039

**Reply by VictorCMiraldo** [reply]
Indeed, but not quite. There's also certainly an issue with the haskell-language-server, because this happens only in a few specific projects. :/

Nevertheless, having access to plaintext docs would be an amazing workaround in the meanwhile!

**Reply by joaotavora** [reply]
> Indeed, but not quite. 

Yes, I agree. It seems that _here_, Emacs and `markdown-mode` are struggling to fontify a temporary buffer  with "github flavoured markdown", an operation that shouldn't really take very long.  Since you can interrupt the Emacs with the `SIGUSR2` signal, you can probably operate Emacs and switch to that temporary buffer, and see its contents.  Could be that it's very very large, and maybe `markdown-mode` could just give up after some time. Or maybe, I'm just guessing, could be that it's quite small but it has some obscure markdown feature that's tripping up `markdown-mode`. Regardless the problem could be isolated there, and then a bug report with a simple no-LSP-server-required MRE recipe can be made to either Emacs' or `markdown-mode` 's bug trackers.

**Reply by VictorCMiraldo** [reply]
> Since you can interrupt the Emacs with the SIGUSR2 signal, you can probably operate Emacs and switch to that temporary buffer, and see its contents

That's a good idea! Indeed, I can operate Emacs after SIGUSR2. I'll re-enable `eglot--format-markup` and try to make a MRE this weekend. Thank you!

**Reply by VictorCMiraldo** [reply]
I didn't manage to find a buffer with the information we wanted, but I expanded the `...` abbreviation of the stack frame until about 65k characters long and it contained a huge amount of generated haskell code, which I would really only expect to see if I'm compiling with all sorts of `-ddump` flags enabled to dump intermediate stages.

I can't really paste that entire string here because it contains proprietary code but this is the general look of it (after replacing the `\n` for actual newlines):
````
```haskell
$dShow :: Show DownloadAttempt
```

* * *

```haskell
$dShow :: Show DownloadAttempt
```

* * *

```haskell
$dShow :: Show CacheControlHeaders
```

* * *

```haskell
$dShow :: Show CacheControlHeaders
```

* * *

```haskell
$dShow :: Show DownloadException
```

* * *

```haskell
$dAllNullary :: AllNullary
  (M1
     C
     ('MetaCons \"ResponseCodeException\" 'PrefixI 'False)
     (S1
        ('MetaSel
           'Nothing 'NoSourceUnpackedness 'NoSourceStrictness 'DecidedLazy)
        (Rec0 ResponseCodeError)))
  False
```
````

And that goes on and on for 3k more lines, then I stopped expanding.


I noticed this happens when the cursor comes to an empty line. I have no idea how the documentation-on-hover works, but could `eglot` be asking for info on an empty symbol and the server replying with all it knows?

**Reply by VictorCMiraldo** [reply]
I made a better workaround and it seems to fix the problem and I still get access to documentation:
```lisp
    (advice-add 'eglot--format-markup :around #'limit-argument-length)
    (defun limit-argument-length (orig input)
      (let ((ty (plist-get input :kind))
            (v  (plist-get input :value)))
        (if (string= ty "markdown")
            (funcall orig (list :kind ty :value (seq-take v 700)))
            (funcall orig input))))
```
